### ISSUE TITLE
Reading fastq records - missing records?

### ISSUE THREAD

**Issue by kaktus42** [seqan/seqan#1030]
Following code reads a fastq-file.
The file I'm testing on has 263333413 records:
```
$ zcat C563WACXX_1.fastq.gz | wc -l
1053333652
```

The problem is, that the code seems to only read
the first 4000000 records. I don't understand why.

I removed try/catch for better readability
```c++
seqan::StringSet<seqan::CharString> ids;
seqan::StringSet<seqan::Dna5QString> seqs;
seqan::StringSet<seqan::CharString> quals;
unsigned batchSize = 10000;
do {
seqan::clear(ids); seqan::clear(seqs); seqan::clear(quals);
readRecords(ids, seqs, quals, inputFile, batchSize);
processRecords(ids, seqs, quals, barcodeGraph, vertexFilePtrMap);
totalProcessed += seqan::length(ids);
std::cout << "\r";
std::cout << "reads processed in total: " << totalProcessed;
std::cout.flush();
} while(seqan::length(ids) > 0);
```

**Comment by temehi** [comment]
@kaktus42 I couldn't reproduce the error. I tried your code after commenting the line
`processRecords(ids, seqs, quals, barcodeGraph, vertexFilePtrMap);`
on  a fastq file containing 152,014,183 reads and I was able read them with out a problem. It could be the case that your code under `processRecords(...)` can not handle too many records.

Could you please try commenting the same line in your code and see if you still have the same problem?

**Comment by catkira** [comment]
I experienced that read_record silently skips records where the length value in the id string is not valid. Is this behaviour intended?

**Comment by temehi** [comment]
@catkira do you mean by not valid length?

**Comment by catkira** [comment]
The 51 in this file, if you change it to some other value, the record is skipped.

@SRR1175698.1:TL:NCTGA SEB9BZKS1:84:D1F78ACXX:4:1101:1969:2170 length=51
GATCGGAAGAGCACACGTCTGAACTCCAGTCACTTAGGCATC
+
HGHHJJHIIIIHIIJJJIJJIIJJIJJJJGHIJJJJJJJGHF

**Comment by temehi** [comment]
@catkira  Are you sure. I couldn't reproduce the bug. The content of the id line is not considered at all which means the length=51 doesn't have any special meaning. Probably you introduced a windows line ending when you try to edit the length value as it is at the end of the id line. Here is the file I used for testing
```
@M02025:17:000000000-A5P81:1:1101:15680:1332 1:N:0:1 length 259
CCTACGTGAGGCAGCAGTAAGGGATATTGGTCAATGGGCGGGAGCCTGAACCAGCCATGCCGCGTGAGGG
ATTGAGGCCCTACGGGTCGTAAACCTCTTTTGCCGGGGGATAACGGGCGCCACGCGTGGCGCCGTGAAGG
TACCCGGAGAATAAGCATCGGCTAACTCCGTGCCAGCAGCCGCGTTAATACGGAGGATGCGAGCGTTATC
CTGATTTATTGGGTTTACAGGGTGCGCAGGCGGGCGTCTTA
+
AAAAAA1B>1>11AABBBAABBBBBBDFGHHGFHHCHCGGGGGGGGHGHHFHHHHGHHHHGGGGGGCGHG
FHHEFHGGGGGGEEGGGCCGHFHGHGGHGHHHBGCGGGGGFFHGCG.CGGGGGGGGG@GG@BG@@@EEFF
FF9BFF@?@@FF//BBFB/A-;AB/BFFFBFEF-9:E----;-99-;-///9--A@BFB9-;@=@BAAE/
//9/;B///;9BA-;;;/:BF-B------;------;-9-;
@M02025:17:000000000-A5P81:1:1101:15671:1349 1:N:0:1 length 251
CCTACGGGAGGCAGCAGTAAGGGATATTGGTCAATGGGCGGGAGCCTGAACCAGCCATGCCGCGTGAGGG
ATTGAGGCCCTACGGGTCGTAAACCTCTTTTGCCGGGGGATAACGGGCGCCACGCGTGGCGCCGTGAAGG
TACCCGGAGAATAAGCATCGGCTAACTCCGTGCCAGCTGCCGCGGTAATACGGAGGATGCGAGCGTTATC
CGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGGCGTCTAA
+
>AABAAA>A?22AAABBBBBBBBBBBDGGHHGHHHG3FFEEEEGGGHHHHHHHHHGHHHHHGGGGGGGHG
GHHGGHHGGGHHFDGGGGGGHHHHHHHHHHHHHHGGGGGGGHFH?@DG-9@FGGGGFGGGFFFFFFF?FF
FFFFFD=@DAFFFFFFFFFFB-ADBBFFFFFFFAFFF///;@FFCAA?BB/A;ADFFFFF;DFFF:D?F9
FFFF?.9:FBBF.AEFF/BFFAA:99;--;DC;-:A--AE;
@M02025:17:000000000-A5P81:1:1101:14449:1427 1:N:0:1 length 251
CCTACGGGTGGCAGCAGTAAGGGATATTGGTCAATGGGCGGGAGCCTGAACCAGCCATGCCGCGTGAGGG
AATGAGGCCCTACGGGTCGTAAACCTCTTTTGCCGGGGGATAACGGGCGCCACGGGTGGCGCCGTGAAGG
TACCCGGAGAATAAGCATCGGCTAACTCCGTGCCAGCAGCCGCGGTAATACGGAGGATGCGAGCGTTATC
CGGATATATTGGGTTTAAAGGGTGCGCATGCGGGCGGCTAA
+
>>AAAABB2>24BAABBBBBBB4BBBDFGGHGHHHHHHHGGCGGGGHHHHHHHHHGHHHGFGGGGGFGHG
GHHFHHHGEGHH0DGGGCFGHGHHFHHHHHHHHHGCGGGDGHHHGGG??DGGGGAAAGGGFFFAFA?EFF
FBFF/A>BFCFFFFFFF//A.9AA//;BF.EEFAFFFB.B.9BB-.;AFFBE.;9-.B/.>D=@@9:AE/
/ABD?./BFFFF?FFDB/;BFBA99999.::;-------.;
``` 
The id of the first record has a wrong length but it didn't case a problem at all.

**Comment by esiragusa** [comment]
Let's backlog this until we get further evidence...

**Comment by catkira** [comment]
int main()
{
    SeqFileIn seqFileIn("test.fastq");
    unsigned i(0);
    seqan::CharString id;
    seqan::Dna5QString seq;
    while (!atEnd(seqFileIn))
    {
        seqan::readRecord(id, seq, seqFileIn);
        std::cout << ++i << std::endl;
    }
	return 0;
}

**Comment by catkira** [comment]
@SRR1175698.5 AdapterXX length=51
GGGGGCTGAAGATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCCGA
+SRR1175698.5 SEB9BZKS1:84:D1F78ACXX:4:1101:2363:2237 length=51
CCCFFFFFHHHHHHIIHIIIJHHIIIFHGGFEFF>FGGGGIEEIDHAHGHC
@SRR1175698.1 Adapter1 length=51
NCTGACTGAGATCGGAAGAGCACACGTCCGTCTGAACTCCAGTCACTTAAA
+SRR1175698.1 SEB9BZKS1:84:D1F78ACXX:4:1101:1969:2170 length=51
#1=DDFFFHHGHHJJHIIIIHIIJJJIJJIIJJIJJJJGHIJJJJJJJGHF
@SRR1175698.2 Adapter2 length=51
NGGGTCTGACATTTTGGAGGCACATCTGCGATCGGAAGAGCACACGTCTGAAA
+SRR1175698.2 SEB9BZKS1:84:D1F78ACXX:4:1101:2216:2128 length=51
#1=DADEFGHHHHJJJJHIIGIJJIJJJIJJJDIJJJJGHIIIJIIIFGHI
@SRR1175698.3 Adapter3 length=51
NCAGCCTGAAACAGATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTC
+SRR1175698.3 SEB9BZKS1:84:D1F78ACXX:4:1101:2091:2173 length=51
#1=DDFFFHHHHHIJJIIJJJJJJJEHJJGGIIJJJJJJFHIJIJJGIIJI
@SRR1175698.4 Adapter6 length=51
NCTTGCTGAGCCTTCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCCC
+SRR1175698.4 SEB9BZKS1:84:D1F78ACXX:4:1101:2237:2190 length=51
#4:ADD>?DDD>2AC:EEDE9BBD=B
@SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
NATGTCTGATCCACAAGGTAGTGTCCTCGGCCCCCTGCTCTACTGCCTGTA
+SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
#1:DDDBDDDHHHIIGGEAFCFHHGAGGIBGEHIIGIIFIDBBFHIGHCCE
@SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
CATTTGGGATCCACAAGGTAGTGTCCTCGGCCCCCTGCTCTACTGCCTGTA
+SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
#1:DDDBDDDHHHIIGGEAFCFHHGAGGIBGEHIIGIIFIDBBFHIGHCCE

**Comment by catkira** [comment]
@SRR1175698.5 AdapterXX length=51
GGGGGCTGAAGATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCCGA
+SRR1175698.5 SEB9BZKS1:84:D1F78ACXX:4:1101:2363:2237 length=51
CCCFFFFFHHHHHHIIHIIIJHHIIIFHGGFEFF>FGGGGIEEIDHAHGHC
@SRR1175698.1 Adapter1 length=51
NCTGACTGAGATCGGAAGAGCACACGTCCGTCTGAACTCCAGTCACTTAAA
+SRR1175698.1 SEB9BZKS1:84:D1F78ACXX:4:1101:1969:2170 length=51
#1=DDFFFHHGHHJJHIIIIHIIJJJIJJIIJJIJJJJGHIJJJJJJJGHF
@SRR1175698.2 Adapter2 length=51
NGGGTCTGACATTTTGGAGGCACATCTGCGATCGGAAGAGCACACGTCTGAA
+SRR1175698.2 SEB9BZKS1:84:D1F78ACXX:4:1101:2216:2128 length=51
#1=DADEFGHHHHJJJJHIIGIJJIJJJIJJJDIJJJJGHIIIJIIIFGHI
@SRR1175698.3 Adapter3 length=51
NCAGCCTGAAACAGATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTC
+SRR1175698.3 SEB9BZKS1:84:D1F78ACXX:4:1101:2091:2173 length=51
#1=DDFFFHHHHHIJJIIJJJJJJJEHJJGGIIJJJJJJFHIJIJJGIIJI
@SRR1175698.4 Adapter6 length=51
NCTTGCTGAGCCTTCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCCC
+SRR1175698.4 SEB9BZKS1:84:D1F78ACXX:4:1101:2237:2190 length=51
#4:ADD>?DDD>2AC:EEDE<FDEAC<DCEDD066D@AD98?DB>9BBD=B
@SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
NATGTCTGATCCACAAGGTAGTGTCCTCGGCCCCCTGCTCTACTGCCTGTA
+SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
#1:DDDBDDDHHHIIGGEAFCFHHGAGGIBGEHIIGIIFIDBBFHIGHCCE
@SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
CATTTGGGATCCACAAGGTAGTGTCCTCGGCCCCCTGCTCTACTGCCTGTA
+SRR1175698.6 SEB9BZKS1:84:D1F78ACXX:4:1101:2595:2166 length=51
#1:DDDBDDDHHHIIGGEAFCFHHGAGGIBGEHIIGIIFIDBBFHIGHCCE

**Comment by catkira** [comment]
its a very strange bug, i also dont get this behaviour with every file. But when I change line 10 to 
NGGGTCTGACATTTTGGAGGCACATCTGCGATCGGAAGAGCACACGTCTGA
then I get 7 reads. With the above file I get 6 reads.

**Comment by temehi** [comment]
@catkira That is because the quality line should have equal length with the sequence line. I personally don't see a bug here. I have also checked with your example. And you should get only 5-6 not 6-7 records as the last two records have identical sequence name.

**Comment by catkira** [comment]
why does this file give 2 records?
@SRR1175698.5 AdapterXX length=51
GGGGGCTGAAGATCGGAAGAGCACACGTCTGGATCCACGACGCTCTTCCGA
+SRR1175698.5 SEB9BZKS1:84:D1F78ACXX:4:1101:2363:2237 length=51
CCCFFFFFHHHHHHIIHIIIJHHIIIFHGGFEFF>FGGGGIEEIDHAHGHC
@SRR1175698.1 Adapter1 length=51
NCTGACTGAGATCGGAAGAGCACACGTCCGTCTGAACTCCAGTCACTTAAA
+SRR1175698.1 SEB9BZKS1:84:D1F78ACXX:4:1101:1969:2170 length=51
#1=DDFFFHHGHHJJHIIIIHII

**Comment by temehi** [comment]
hmmm... 
I see what the problem is. The reason for the two records out of your file is that the bad record violating the fastq specs is the last record in the file. 

In your previous example the bad recored was in the middle. what happened was the qualities are read starting from '+' character down to the length of the sequence just read. This will eat-up the '@' character from the next recored as a quality. which leads to skipping the recored next to a bad recored (qualities shorter than sequences).

Usually fastq files should follow the specifications. Never the less seqan could handle this exceptions better.  I will work on it.

Thanks for reporting this issue and staying responsive. 
great observation :+1:

**Comment by h-2** [comment]
fixed by #1299
